storage/memorybysubnet/peer_store.go

```diff
@@ -0,0 +1,595 @@
+	prometheus.MustRegister(promInfohashesCount)
+
+	// Register the storage driver.
+	storage.RegisterDriver("memorybysubnet", driver{})
+}
+
+var promGCDurationMilliseconds = prometheus.NewHistogram(prometheus.HistogramOpts{
+	Name:    "chihaya_storage_gc_duration_milliseconds",
+	Help:    "The time it takes to perform storage garbage collection",
+	Buckets: prometheus.ExponentialBuckets(9.375, 2, 10),
+})
+
+var promInfohashesCount = prometheus.NewGauge(prometheus.GaugeOpts{
+	Name: "chihaya_storage_infohashes_count",
+	Help: "The number of Infohashes tracked",
+})
+
+// recordGCDuration records the duration of a GC sweep.
+func recordGCDuration(duration time.Duration) {
+	promGCDurationMilliseconds.Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+// recordInfohashesDelta records a change in the number of Infohashes tracked.
+func recordInfohashesDelta(delta float64) {
+	promInfohashesCount.Add(delta)
+}
+
+type driver struct{}
+
+func (d driver) NewPeerStore(icfg interface{}) (storage.PeerStore, error) {
+	// Marshal the config back into bytes.
+	bytes, err := yaml.Marshal(icfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the bytes into the proper config type.
+	var cfg Config
+	err = yaml.Unmarshal(bytes, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return New(cfg)
+}
+
+// ErrInvalidGCInterval is returned for a GarbageCollectionInterval that is
+// less than or equal to zero.
+var ErrInvalidGCInterval = errors.New("invalid garbage collection interval")
+
+// Config holds the configuration of a memory PeerStore.
+type Config struct {
+	GarbageCollectionInterval   time.Duration `yaml:"gc_interval"`
+	PeerLifetime                time.Duration `yaml:"peer_lifetime"`
+	ShardCount                  int           `yaml:"shard_count"`
+	PreferredIPv4SubnetMaskBits int           `yaml:"preferred_ipv4_subnet_mask_bits"`
+	PreferredIPv6SubnetMaskBits int           `yaml:"preferred_ipv6_subnet_mask_bits"`
+}
+
+// New creates a new PeerStore backed by memory.
+func New(cfg Config) (storage.PeerStore, error) {
+	shardCount := 1
+	if cfg.ShardCount > 0 {
+		shardCount = cfg.ShardCount
+	}
+
+	if cfg.GarbageCollectionInterval <= 0 {
+		return nil, ErrInvalidGCInterval
+	}
+
+	ps := &peerStore{
+		shards:   make([]*peerShard, shardCount*2),
+		closed:   make(chan struct{}),
+		ipv4Mask: cfg.PreferredIPv4SubnetMaskBits,
+		ipv6Mask: cfg.PreferredIPv6SubnetMaskBits,
+	}
+
+	for i := 0; i < shardCount*2; i++ {
+		ps.shards[i] = &peerShard{swarms: make(map[bittorrent.InfoHash]swarm)}
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ps.closed:
+				return
+			case <-time.After(cfg.GarbageCollectionInterval):
+				before := time.Now().Add(-cfg.PeerLifetime)
+				log.Debugln("memory: purging peers with no announces since", before)
+				ps.collectGarbage(before)
+			}
+		}
+	}()
+
+	return ps, nil
+}
+
+type serializedPeer string
+
+type peerShard struct {
+	swarms map[bittorrent.InfoHash]swarm
```

Comment:
We should make some assumptions about the number of infohashes, peers and subnets. Based on that we might want to have a `map[subnet]map[InfoHash]swarm` here to save the amount of maps (because they use a lot of memory).